adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java

```diff
@@ -195,80 +192,356 @@ public void run() {
+         */
+        private static final String REGEX_SPLIT_UNQUOTED_COMMA = "," + REGEX_UNQUOTED_LOOKAHEAD;
+
+        /**
+         * Regex sequence to parse schemes from WWW-Authenticate header values.
+         */
+        private static final String REGEX_STRING_TOKEN_WITH_SCHEME = "^([^\\s|^=]+)[\\s|\\t]+([^=]*=[^=]*)+$";
+
+        /**
+         * Comma+space suffix used to preserve formatting during parsing.
+         */
+        private static final String SUFFIX_COMMA = ", ";
+
+        /**
+         * The authentication scheme of this challenge (ex. Basic, Bearer).
+         */
+        private String mScheme;
+
+        /**
+         * The parameters of this scheme (ex. realm, authoritization_uri, scope).
+         */
+        private Map<String, String> mParameters;
+
+        /**
+         * Constructs a new Challenge.
+         *
+         * @param scheme The scheme used by this challenge.
+         * @param params The params which accompany this challenge.
+         */
+        private Challenge(final String scheme, final Map<String, String> params) {
+            mScheme = scheme;
+            mParameters = params;
+        }
+
+        /**
+         * Gets the scheme.
+         *
+         * @return The scheme to get.
+         */
+        public String getScheme() {
+            return mScheme;
+        }
+
+        /**
+         * Gets the parameters.
+         *
+         * @return The parameters to get.
+         */
+        public Map<String, String> getParameters() {
+            return mParameters;
+        }
+
+        /**
+         * Parses a single challenge String, typically the value field of a WWW-Authenticate header.
+         *
+         * @param challenge The challenge String to parse.
+         * @return The Challenge object derived from challenge.
+         * @throws ResourceAuthenticationChallengeException If a parsing error is encountered or
+         *                                                  the String is malformed.
+         */
+        static Challenge parseChallenge(final String challenge) throws ResourceAuthenticationChallengeException {
+            final String scheme = parseScheme(challenge);
+            Logger.d(TAG, "Parsing scheme: " + scheme);
+            final String challengeSansScheme = challenge.substring(scheme.length() + 1);
+            final Map<String, String> params = parseParams(challengeSansScheme);
+            return new Challenge(scheme, params);
+        }
+
+        /**
+         * Parses the scheme of a challenge String.
+         *
+         * @param challenge The challenge String to parse.
+         * @return The scheme portion of the challenge String.
+         * @throws ResourceAuthenticationChallengeException If a parsing error is encountered or
+         *                                                  the String is malformed.
+         */
+        private static String parseScheme(String challenge) throws ResourceAuthenticationChallengeException {
```

Comment:
So now it will accept all kinds of scheme, right? Not just "BEARER"?

Reply:
Correct, any scheme can now be used/parsed